firmware/ethereum.c

```diff
@@ -0,0 +1,551 @@
+			fsm_sendFailure(FailureType_Failure_Other, "Data length provided, but no initial chunk");
+			ethereum_signing_abort();
+			return;
+		}
+		/* Our encoding only supports transactions up to 2^24 bytes.  To
+		 * prevent exceeding the limit we use a stricter limit on data length.
+		 */
+		if (msg->data_length > 16000000)  {
+			fsm_sendFailure(FailureType_Failure_Other, "Data length exceeds limit");
+			ethereum_signing_abort();
+			return;
+		}
+		data_total = msg->data_length;
+	} else {
+		data_total = 0;
+	}
+	if (msg->data_initial_chunk.size > data_total) {
+		fsm_sendFailure(FailureType_Failure_Other, "Invalid size of initial chunk");
+		ethereum_signing_abort();
+		return;
+	}
+
+	// safety checks
+	if (!ethereum_signing_check(msg)) {
+		fsm_sendFailure(FailureType_Failure_ActionCancelled, "Signing aborted (safety check failed)");
+		ethereum_signing_abort();
+		return;
+	}
+
+	layoutEthereumConfirmTx(msg->to.bytes, msg->to.size, msg->value.bytes, msg->value.size);
+	if (!protectButton(ButtonRequestType_ButtonRequest_SignTx, false)) {
+		fsm_sendFailure(FailureType_Failure_ActionCancelled, "Signing cancelled by user");
+		ethereum_signing_abort();
+		return;
+	}
+
+	if (data_total > 0) {
+		layoutEthereumData(msg->data_initial_chunk.bytes, msg->data_initial_chunk.size, data_total);
+		if (!protectButton(ButtonRequestType_ButtonRequest_SignTx, false)) {
+			fsm_sendFailure(FailureType_Failure_ActionCancelled, "Signing cancelled by user");
+			ethereum_signing_abort();
+			return;
+		}
+	}
+
+	layoutEthereumFee(msg->value.bytes, msg->value.size,
+					  msg->gas_price.bytes, msg->gas_price.size,
+					  msg->gas_limit.bytes, msg->gas_limit.size);
+	if (!protectButton(ButtonRequestType_ButtonRequest_SignTx, false)) {
+		fsm_sendFailure(FailureType_Failure_ActionCancelled, "Signing cancelled by user");
+		ethereum_signing_abort();
+		return;
+	}
+	
+	/* Stage 1: Calculate total RLP length */
+	uint32_t rlp_length = 0;
+
+	layoutProgress("Signing", 0);
+
+	rlp_length += rlp_calculate_length(msg->nonce.size, msg->nonce.bytes[0]);
+	rlp_length += rlp_calculate_length(msg->gas_price.size, msg->gas_price.bytes[0]);
+	rlp_length += rlp_calculate_length(msg->gas_limit.size, msg->gas_limit.bytes[0]);
+	rlp_length += rlp_calculate_length(msg->to.size, msg->to.bytes[0]);
+	rlp_length += rlp_calculate_length(msg->value.size, msg->value.bytes[0]);
+	rlp_length += rlp_calculate_length(data_total, msg->data_initial_chunk.bytes[0]);
+
+	/* Stage 2: Store header fields */
+	hash_rlp_list_length(rlp_length);
+
+	layoutProgress("Signing", 100);
+
+	hash_rlp_field(msg->nonce.bytes, msg->nonce.size);
+	hash_rlp_field(msg->gas_price.bytes, msg->gas_price.size);
+	hash_rlp_field(msg->gas_limit.bytes, msg->gas_limit.size);
+	hash_rlp_field(msg->to.bytes, msg->to.size);
+	hash_rlp_field(msg->value.bytes, msg->value.size);
+	hash_rlp_length(data_total, msg->data_initial_chunk.bytes[0]);
+	if (data_total > 1 || msg->data_initial_chunk.bytes[0] >= 0x80) {
```

Comment:
I'm wondering if it is better if hash_rlp_length hashes nothing for length==1 and initial_byte < 0x80.  Then we can remove the `if` here and in hash_rlp_field.